### PR TITLE
[WIP] flat source foil geometry. 

### PR DESCRIPTION
**Pull request by lemiere**
2 features : new geometry and new vertex generator

* Issue with variant system due to new resources files organization
* Foils are not well organized for now (ideal --> real TBD)

--> Need help to get the new geometry available as a variant

### CODE REVIEW

**Comment by drbenmorgan** [comment]
Thanks @lemiere! The tests provide a good overview of the errors, so I'll take a look at these. Is the main issue just how to add the `realistic` variant?

I've added a few more relevant people for any input/feedback/questions they may have.

**Comment by lemiere** [comment]
For now, there is only variant issue but I did not test more ....

**Comment by cherylepatrick** [comment]
Fantastic! Let me know when the vertex generator is available and I will organise testing to check it matches Andrea's document. Question - how do we test the variant geometry is properly implemented? (I understand how to test the vertex generator).

**Comment by lemiere** [comment]
Alone in the dark, I am ! Before the publication, long the road will be....
Testing the geometry variant you will, by using the bxgeom_inspector my apprentice !

**Comment by cherylepatrick** [comment]
Oh I can do that, I saw a fantastic tutorial about it if I recall correctly...

**Comment by drbenmorgan** [comment]
Hi @lemiere, just starting to look at this, and am reproducing the errors seen in testing. The primary one I think is simple, the section here:

https://github.com/SuperNEMO-DBD/Falaise/blob/beda690363e06a05414de6a1dfe76c1f915d8baa/resources/snemo/demonstrator/geometry/GeometryManager.conf#L39-L50

probably needs to be in a `variant_{if,endif}` block as I _think_ the `variant_only` directive only applies to one line. However, I don't yet see where a conditional on the type of source foils would come in within `GeometryManager.conf`

Edit: Looking through further, it looks like `GeometryManager.conf` _shouldn't_ need any specialisation for the `source_layout` variant. Geometry's only use of conditionals on `source_layout` is in `source_blocks.geom`. So I _think_ changes to `GeometryManager.conf` can be reverted, but I may well be missing something here.

File: programs/flreconstruct/CMakeLists.txt [hunk 1]
```diff
@@ -25,6 +25,14 @@
 # - Bit hacky for now
 find_package(Boost 1.60 REQUIRED program_options)
 
+
+
+find_package (Qt5Core)
+find_package (Qt5Widgets)
+find_package (Qt5PrintSupport)
```
Comment (drbenmorgan) [inline programs/flreconstruct/CMakeLists.txt:32]: Why are these lines needed? The top level `find_package` on Bayeux should refind the Qt5 dependencies.

File: programs/flsimulate/CMakeLists.txt [hunk 2]
```diff
@@ -26,6 +26,13 @@
 find_package(Boost 1.60 REQUIRED program_options filesystem system)
 find_package(Geant4 REQUIRED)
 
+
+find_package (Qt5Core)
+find_package (Qt5Widgets)
+find_package (Qt5PrintSupport)
```
Comment (drbenmorgan) [inline programs/flsimulate/CMakeLists.txt:32]: As above, the top level `find_package(Bayeux)` should find these.

File: programs/flvisualize/CMakeLists.txt [hunk 3]
```diff
@@ -17,6 +17,15 @@ project(FLVisualize VERSION 3.0.0)
 set(Boost_COMPONENTS thread program_options)
 find_package(Boost 1.55 REQUIRED ${Boost_COMPONENTS})
 
+
+# - Qt5 support
+find_package (Qt5Core)
+find_package (Qt5Widgets)
+find_package (Qt5PrintSupport)
+
```
Comment (drbenmorgan) [inline programs/flvisualize/CMakeLists.txt:25]: As above...

**Comment by lemiere** [comment]
At flsimulate exec, I got : 
`[void materials::manager::load(const datatools::multi_properties&):425: Missing property 'material' for a material alias !]
`
Even if materials used in : 
`GeometryModels/source_module/flat/source_foils.geom` 
are defined into 
 `materials/Materials.conf`

Next day, next step.... I hope

**Comment by cherylepatrick** [comment]
OK, so I think I finally understand what all the bits of this are - gosh it is complicated!

But what I don't understand is why it won't run. What's changed, if anything, in the resource file formats? What's our next step for debugging it? It's an absolutely enormous chuck of work - is our best plan to try breaking it up into smaller chunks and adding it bit by bit? Can we do that without having to rebuild falaise each time, or does that need #205 ? Can we test by adding the resource files a bit at a time into the #205 branch, if we aren't ready to merge it? *still not very good at GitHub skillz*

**Comment by drbenmorgan** [comment]
> But what I don't understand is why it won't run. What's changed, if anything, in the resource file formats? 

*Nothing*! The layout has changed to remove the "tag" level, and a few things were moved into more logical locations. If that was the issue, we'd be seeing errors related to "file not found" or similar. The original PR using "tags" is here: #113, but it's difficult to separate the tag management part from the parameter changes.

> Is our best plan to try breaking it up into smaller chunks and adding it bit by bit? 

I'd agree with this - get the new materials in first. Then the variants, purely to check that we can switch between. Then the detailed geometry.

> Can we do that without having to rebuild falaise each time, or does that need #205 ?

I'd strongly recommend getting #205 merged and then rebase this on to the updated `develop`. This will guarantee that edits to the resource files are reflected immediately in the tests/direct runs of `flsimulate`.

> Can we test by adding the resource files a bit at a time into the #205 branch, if we aren't ready to merge it?

No - #205 is orthogonal to this. Get #205 merged, then rebase this one.

If multiple people want to work on this PR, then there's a guide on how to do this: https://github.com/SuperNEMO-DBD/Falaise/blob/develop/CONTRIBUTING.md#working-with-others

It just requires the owner of the branch to grant push access to the relevant branch for any co-workers (or you can use Pull Requests as outlined in the docs).

Bottom line though, the issues here are down to the lack of documentation for Bayeux, especially geomtools, variants, and materials. We should discuss this in the next Software Integration meeting, as this PR's an ideal opportunity to address this.

**Comment by drbenmorgan** [comment]
Now that #205's merged, I suggest @lemiere rebases the `flat_source_foils` branch onto the updated `develop` branch and force push it. It'll trigger a new build which I expect will fail, but gets up into a position to properly start testing with better guarantees about editing and testing the resources.

**Comment by cherylepatrick** [comment]
That sounds like a good idea to me. Then hopefully we can get someone else (Malak, maybe, as she has kindly agreed to test this) to work on testing the changes step by step. I can help. @lemiere - are you OK with that plan?

**Comment by lemiere** [comment]
Yes sure ! 
I've to find the right git cmd to proceed, then I will do it

**Comment by cherylepatrick** [comment]
> Yes sure !
> I've to find the right git cmd to proceed, then I will do it

Haha I know that feeling! Sigh!

**Comment by drbenmorgan** [comment]
It should be, assuming you have a remote named "upstream" that points to here, then:

```
$ git fetch upstream
$ git checkout flat_source_foils
$ git rebase upstream/develop
```

See https://github.com/SuperNEMO-DBD/Falaise/blob/develop/CONTRIBUTING.md#develop-code for more details.

**Comment by cherylepatrick** [comment]
Thank you @drbenmorgan !

**Comment by cherylepatrick** [comment]
Re: documentation - @lemiere has pointed me to some docs that were in the old Falaise 3.2.0 installation - for example /resources/config/snemo/demonstrator/geometry/4.0/README.rst and equivalent for the other folders in the resources section.

They seem to have gotten lost in later versions but they have good info in them so we should reinstate them -  but I don't know if the best place is in the same spots they were in before (you're already half-way there when you've found the right folder!). @drbenmorgan - do you think they could be included in the online Falaise docs? I also thought that I was hoping we could ask for help from the students in tracking down this issue, that I should probably write some tutorial-like explanation, so we could add that too. What do you think?

**Comment by fmauger** [comment]
Hi all
I have fixed the config files in my own branch (because I was not able to sanely import Yves' PR #197 in my own environment with three waiting branches fixing various odd things). It seems to work now with the  "RealSource" source layout which is accessible through the variant system as an alternative to the legacy "Basic" source layout. I have reorganized the files originally proposed by Yves but basically everything was there but some subtle scratch concerning material definition depending on the variant options.
I also added a README file in the geometry config dir to explain how to browse the geometry and test it for a new feature.
I have to work a little bit more on it to make sure I did not miss something.
Should I provide also some associated vertex generators within the same thread or would you prefer to have a specific PR for that ? This would mean that no vertex generation would be possible from this new layout until the vertex gen PR is accepted.

snemo-real_source_foils-1

**Comment by drbenmorgan** [comment]
Instructions on how to collaborate on a branch are here: https://github.com/SuperNEMO-DBD/Falaise/blob/develop/CONTRIBUTING.md#working-with-others

Let me know if these need any clarification or extensions! You can also add people as "developers" on your fork to allow direct push access (but make sure you coordinate carefully to avoid issues).

**Comment by cherylepatrick** [comment]
Thank you LOTS @fmauger - that's amazing. I vote for adding the vertex generators, then we can start testing. Malak has kindly offered to test for us, so could you add her to the fork. Thanks for adding the docs too - I'm happy to read through them if you point me at the fork also.

**Comment by fmauger** [comment]
Ok let's implement the VGs! 
Yves has already provided most of the definition files.
Below is a first result using a vertex generator in the bulk of a specific pad in a specific strip:
snemo-realistic_flat_source_foils-vertex_23_3-1
I have renamed this geometry variant as "geometry:layout/if_basic/source_layout/if_realistic_flat"
in addition to the legacy "geometry:layout/if_basic/source_layout/if_basic" we have used before.
Also the variant generators have been grouped in new "SourceFoilRealisticFlat" group to ease the lisibility within the GUI (menu list with soooo many items). 
Now I have to add the combined VGs calculated by Yves for the generation on the full source foil bulk or surface then check through flsimulate with a specific validation test.
To be continued.

**Comment by fmauger** [comment]
Also note that Yves will next modify this proposal to take into account the permutations of source strips done by Andrea before the detector closing. This won't change the interface, just the position of geometry model in the frame.

**Comment by cherylepatrick** [comment]
Hi @lemiere @fmauger @pfranchini - we've let this slip for a while but I'd love to see it sorted out by the collaboration meeting. If I understand correctly, the next step is to reorder the source foils. Is that difficult to do? We have the ordering available in Andrea's document. Does someone know how to progress with this? Then the next step will be to work out who might be able to do it...

**Comment by drbenmorgan** [comment]
Bumping again - @lemiere, @fmauger can we have an update on the status and estimated timeline for completing this **by the end of this week** please?

**Comment by fmauger** [comment]
Hi

A first version of the realistic layout of the source foils is waiting here:
https://github.com/fmauger/Falaise/tree/feature-flat_source_foils/
with the proper geometry and associated additional vertex generators.

However, Yves must now use this preliminary work to reordered the source strips to take into account the final positioning in the source frame (Andrea's final choice to avoid collision of deformed neighbour strips).

**Comment by cherylepatrick** [comment]
Fab, thank you @fmauger ! Yes, Malak has tested this version, I think, and she managed to identify the different types of foils - as you say, they just need reordering. @lemiere - if you need help, I believe @pfranchini has some time, so feel free to contact him if he can be helpful!

**Comment by pfranchini** [comment]
Happy to help, was trying to get up to speed with the issue.

**Comment by lemiere** [comment]
Resume in PR #228
